File: 290-word-pattern/word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] word = s.split(" ");
        if(pattern.length() != word.length) {
            return false;
        }

        Map<Character, String> map1 = new HashMap<>();
        Map<String, Character> map2 = new HashMap<>();

        for(int i = 0; i < pattern.length(); i++) {
            char patChar = pattern.charAt(i);
            if(map1.containsKey(patChar) && !map1.get(patChar).equals(word[i])) {
                return false;
            }
            map1.put(patChar, word[i]);
        }

        for(int i = 0; i < word.length; i++) {
            char patChar = pattern.charAt(i);
            if(map2.containsKey(word[i]) && map2.get(word[i]) != patChar){
                return false;
            }
            map2.put(word[i], patChar);
        }

        return true;
    }
}

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        word = s.split(" ")

        if len(pattern) != len(word):
            return False
        
        seen = {}

        # zip은 이렇게 생각하면 돼요: "abba" + "dog cat cat dog" → (a,dog), (b,cat), (b,cat), (a,dog)

        for p, w in zip(pattern, word):
            key_p = ("p", p)
            key_w = ("w", w)

            if key_p in seen and seen[key_p] != w:
                return False
            if key_w in seen and seen[key_w] != p:
                return False
            seen[key_p] = w
            seen[key_w] = p
        return True

